### DISCUSSION TITLE
11.0.0-preview3

### DISCUSSION THREAD

**Discussion by danwalmsley** [AvaloniaUI/Avalonia#9294] [General]
- stable browser-wasm target.
- stable ios / android

<hr /><em>This discussion was created from the release <a href='https://github.com/AvaloniaUI/Avalonia/releases/tag/11.0.0-preview3'>11.0.0-preview3</a>.</em>

**Comment by pereg-1985** [comment]
Hello. I am informing you about a serious bug that exists in Avalonia starting from 11.0.0-preview-2 version.

When creating a publish version of the project on Avalonia, the finished program starts, hanging for a couple of seconds and breaks. 

Everything is fine with the debug and release versions of the project. But it is impossible to make a ready release of the program.
I use version 11.0.0-preview-1, I can't update it. The problem is not solved.

https://user-images.githubusercontent.com/34482890/198836869-1ef5531b-4caf-4950-81d2-68c397b68a85.mp4

**Reply by maxkatz6** [reply]
Please open an issue with more details and how you build the project.

**Comment by pereg-1985** [comment]
Details in this video:

https://user-images.githubusercontent.com/34482890/198842153-850464be-5903-41b1-89a5-4ea3ee6a7ef3.mp4

**Reply by BAndysc** [reply]
Try to remove obj/ and bin/ folders, I had similar problems in the past when upgrading Avalonia, it wasn't rebuilding everything properly resulting in something similar.

**Reply by maxkatz6** [reply]
RuntimeIdentifier must be win7-x64, not win-x64.

**Comment by pereg-1985** [comment]
Thank you, it works.

**Comment by efef77** [comment]
Issue regarding ICustomDrawOperation:

When porting a Skia custom rendering application using "ICustomDrawOperation" from version 0.10.18 to 11.0.0 prevew-3 i had to do a lot of reflection to gain the SKCanvas which was hidden behind private members like this.

```
public void Render(IDrawingContextImpl context) {
    // Avalonia 0.10.18 method
    //var canvas = (context as ISkiaDrawingContextImpl)?.SkCanvas;

    // Avalonia 11.0.0 preview method
    var fi = context.GetType().GetField("_impl",
        BindingFlags.NonPublic | BindingFlags.Instance);
    var impl = (Avalonia.Platform.IDrawingContextImpl)fi.GetValue(context);
    var pi = impl.GetType().GetProperty("Canvas",
        BindingFlags.Public | BindingFlags.Instance);
    var canvas = (SkiaSharp.SKCanvas)pi.GetValue(impl);

    if (canvas != null) {
        RenderAction?.Invoke(canvas);
    }
}
```

Perhaps this was not the best method or otherwise the final version should allow a better public access to the SKCanvas.
May be some one can pick this up? Thank you.

**Reply by maxkatz6** [reply]
Check this API
https://github.com/AvaloniaUI/Avalonia/pull/8656

**Reply by efef77** [reply]
Thank you maxkatz6,

it looked good but did not work. The code below passed the compiler so it looked promising:

```
var skia = context.GetFeature<ISkiaSharpApiLeaseFeature>();
SKCanvas canvas2 = skia.Lease().SkCanvas;

```
At runtime i gained the very same SkCanvas object but it did not render. So the canvas object gets prepared not the same way but i coud not find out the difference. Regards efef77.

**Reply by maxkatz6** [reply]
> SKCanvas canvas2 = skia.Lease().SkCanvas;

Please follow the "using" patter with disposable objects.
This canvas is "alive" and can be used only inside of the Render method.

**Reply by efef77** [reply]
Hi maxkatz6, 
it worked disposing the Lease (not the canvas) like this:
```
public void Render(IDrawingContextImpl context) {
    // Avalonia 11.0.0 preview feature method
    var skia = context.GetFeature<ISkiaSharpApiLeaseFeature>();
    using (var lease = skia.Lease()) {
        SKCanvas canvas = lease.SkCanvas;
        if (canvas != null) RenderAction?.Invoke(canvas);
    }
}
```
Thank you for helping.

**Comment by lukaszossowski** [comment]
Is there any documentation on the new AvaloniaMainActivity and AvaloniaSplashActivity? Upgrading the packages to preview 3 informs me that there no longer is an AvaloniaActivity which breaks the project.

**Reply by maxkatz6** [reply]
Check new xplat templates, they were updated
